game.py

```diff
@@ -0,0 +1,908 @@
+from math import inf
+from copy import copy, deepcopy
+import random
+import pickle
+import time
+
+class GameState():
+    
+    def __init__(self, board, player_next, connect_number, number_of_moves, moves_list = []):

```

Comment:
The default argument of a list here is sadly no good, standard practice is to default to `None` and then have `if moves_list is None: move_list = []` below. Python generally is pretty good at not allowing you to easily shoot yourself in the foot, but this is one common and glaring exception(!)

https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments